### PR TITLE
percentile and nanpercentile

### PR DESCRIPTION
**Pull request by rjenc29**
Implementation of np.percentile and np.nanpercentile.  
Submitting for CI prior to review request.

### CODE REVIEW

**Comment by rjenc29** [comment]
Hi, please could you review this PR when you have a mo.

The intention is to support np.percentile and np.nanpercentile (first 2 args).  I have used the high-level extension APIs and followed the approach used for median to some extent.

The behaviour of np.percentile was changed in 1.10 with regards to handling arrays which contain NaN.  The behaviour of np.nanpercentile was changed in 1.11 with regards to handling all NaN arrays.  I have dropped comments in the code and have used numpy version gates rather that conditionally branch and replicate the deprecated behaviour.

The performance versus numpy exhibits similar characteristics to median and nanmedian, on my machine.

Thanks!

File: numba/tests/test_array_reductions.py [hunk 1]
```diff
@@ -330,25 +335,61 @@ def variations(a):
...
+        a[a % 4 >= 1] = 3.5
+        yield a
+        a[a % 4 >= 2] = np.nan
+        yield a
+        a[:] = np.nan
+        yield a
+
+    @staticmethod
+    def _percentile_variations(q):
+        # array, list of reversed values, one value, one value as int
```
[the remark this answers is not in the document]
Comment (rjenc29) [reply numba/tests/test_array_reductions.py]: Done

File: numba/tests/test_array_reductions.py [hunk 2]
```diff
@@ -330,25 +335,61 @@ def variations(a):
 
         self.check_median_basic(pyfunc, variations)
 
+    def check_percentile_basic(self, pyfunc, array_variations, percentile_variations, abs_tol=1e-12):
```
[the remark this answers is not in the document]
Comment (rjenc29) [reply numba/tests/test_array_reductions.py]: Added some tests and explicit handling for single element arrays.  

The cases were array includes one or more +/-np.inf are problematic - I find the Numpy behaviour a little difficult to reason about.  Added some commented-out test cases & Numpy outputs for now; will have a think.

The abs_tol: has, yes - I still need to look into this; differences exceed machine eps.

File: numba/targets/arraymath.py [hunk 3]
```diff
@@ -765,6 +765,106 @@ def median_impl(arry):
...
+            return _collect_percentiles(a, percentiles, skip_nan=True)[0]
+
+        def np_nanpercentile_q_seq_impl(a, q):
+            percentiles = np.array(q)
+            return _collect_percentiles(a, percentiles, skip_nan=True)
+
+        def np_nanpercentile_q_array_impl(a, q):
+            return _collect_percentiles(a, q, skip_nan=True)
+
+        if isinstance(q, (types.Float, types.Integer)):
```
[the remark this answers is not in the document]
Comment (rjenc29) [reply numba/targets/arraymath.py]: Done

File: numba/targets/arraymath.py [hunk 4]
```diff
@@ -765,6 +765,106 @@ def median_impl(arry):
...
+
+if numpy_version >= (1, 11):
+    @overload(np.nanpercentile)
+    def np_nanpercentile(a, q):
+
+        # Note: np.nanpercentile was implemented in 1.9, but the return type in the
+        # case of an all-NaN slice was np.nan until 1.11 when it was changed to be
+        # an array of np.NaN of length equal to q.  The 1.9 / 1.10 behaviour is
+        # problematic from a return type unification perspective, so for now using
+        # a 1.11 feature guard, pending review.
```
[the remark this answers is not in the document]
Comment (rjenc29) [reply numba/targets/arraymath.py]: Done

File: numba/targets/arraymath.py [hunk 5]
```diff
@@ -765,6 +765,106 @@ def median_impl(arry):
...
+
+if numpy_version >= (1, 10):
+    @overload(np.percentile)
+    def np_percentile(a, q):
+
+        # Note: np.percentile behaviour in the case of an array containing one or
+        # more NaNs was changed in numpy 1.10 to return an array of np.NaN of
+        # length equal to q.  It might be possible to branch logic based on numpy
+        # version and replicate the 1.9 behaviour, but for now using a 1.10 feature
+        # guard, pending review.
```
[the remark this answers is not in the document]
Comment (rjenc29) [reply numba/targets/arraymath.py]: Done

File: numba/targets/arraymath.py [hunk 6]
```diff
@@ -765,6 +765,106 @@ def median_impl(arry):
 
     return median_impl
 
+@register_jitable
+def _collect_percentiles(a, q, skip_nan=False):
+
+    # check all requested percentiles, q, are in required range
+    assert len(q[np.isnan(q)]) == 0, 'Percentiles must be in the range [0,100]'
```
[the remark this answers is not in the document]
Comment (rjenc29) [reply numba/targets/arraymath.py]: Done

File: numba/tests/test_array_reductions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rjenc29) [reply numba/tests/test_array_reductions.py]: Switching to np.true_divide allows abs_tol to be set to eps.

[the remark this answers is not in the document]
Comment (rjenc29) [reply numba/tests/test_array_reductions.py]: I have tried some heuristics to reproduce NumPy behaviour when processing an array with infinite values, but without much success - the Numpy behaviour seems inconsistent and unreasonable.

**Comment by rjenc29** [comment]
Hi, I believe all review feedback items are now addressed, excepting the handling of arrays which contain infinite values - this has me stumped.  In and of itself, it shouldn't be a big deal but NumPy's behaviour is inconsistent at best.

Not sure on the best way forward - open to suggestion, cheers.

**Comment by stuartarchibald** [comment]
Thanks for the fixes, much appreciated.

Inf/NaN behaviour is hard to trace. However, I am of the view that if possible `inf` input probably should not turn into `nan` (cf. NumPy). This line:
```python
val = lower + m * (upper - lower)
```
is likely the cause as when `lower = inf` and `upper` is finite this leads to `+inf - inf = nan`. If you reorder the computation as:
```python
val = lower * (1 - m) + upper * m
```
it avoids such issues and is a closer approximation to the method NumPy employs. Making this change and testing with:
```python
infnans = [
    np.array([1., 1]),
    np.array([1., np.inf]),
    np.array([np.inf, np.inf]),
    np.array([-np.inf, np.inf]),
    np.array([-np.inf, -np.inf]),
    np.array([1., np.inf, np.inf]),
    np.array([1., -np.inf, np.inf]),
    np.array([1., -np.inf, -np.inf]),
    np.array([1., np.nan]),
    np.array([np.nan, np.nan]),
    np.array([-np.nan, np.nan]),
    np.array([-np.nan, -np.nan]),
    np.array([1., np.nan, np.nan]),
    np.array([1., -np.nan, np.nan]),
    np.array([1., -np.nan, -np.nan]),
    ]
```
gave parity to NumPy when testing locally. Thoughts?

File: numba/targets/arraymath.py [hunk 7]
```diff
@@ -765,6 +765,109 @@ def median_impl(arry):
...
+
+        for i in range(len(q)):
+            percentile = q[i]
+            if percentile == 100:
+                val = np.max(temp_arry)  # bypass pivoting
+            elif percentile == 0:
+                val = np.min(temp_arry)  # bypass pivoting
+            else:
+                rank = 1 + (n - 1) * np.true_divide(percentile, 100.0)  # linear interp between closest ranks
+                f = math.floor(rank)
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: Out of interest, why `math.floor` and not `np.floor`?

Comment (rjenc29) [reply numba/targets/arraymath.py]: I noticed that math.floor is used elsewhere in arraymath.py - I thought I would follow suit, but happy to change.

**Comment by rjenc29** [comment]
oh actually, hold that thought - the heuristics still have have some inconsistency vs NumPy.  I will revisit when back from hols...

**Comment by stuartarchibald** [comment]
I think what you have written does the job, thanks :) Whilst it's a little convoluted I don't think it can easily be helped, it is also on less well visited branches so is likely to have minimal impact. I would perhaps make it such that `num_pos_inf`, `num_neg_inf` and `num_finite` are only computed if needed as the `else` case (`# linear interp between closest ranks`), which is far more common, doesn't need them. Thanks.

**Comment by rjenc29** [comment]
Good stuff; will tidy up over the weekend and commit, cheers.

**Comment by stuartarchibald** [comment]
@rjenc29 Thanks for the latest fix-ups and your persistence in resolving the issues/getting this to work. This is a good contribution, happy to merge.

**Pull request merged**